src/Voron/Data/PostingLists/PForDecoder.cs

```diff
@@ -156,9 +174,15 @@ private static int HorizontalOr(in Vector128<int> x)
         }        
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private unsafe static ulong Read(int stateBitPos, byte* inputBufferPtr, int bitsToRead, out int outputStateBit)
+        private unsafe static ulong Read(int stateBitPos, byte* inputBufferPtr, int bitsToRead, int inputBufferSize, out int outputStateBit)
         {
-            if ( Avx2.IsSupported)
+            var maxByteOffset = (stateBitPos + bitsToRead) / 8;
+#if DEBUG
+            if (maxByteOffset >= inputBufferSize)
+                throw new ArgumentOutOfRangeException();
+#endif
+            //TODO: Ensure that we aren't reading beyond the end of the buffer!
```

Comment:
@redknightlois Note that this need a proper fix.

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-20295